psl/psl-core/src/lib.rs

```diff
@@ -52,14 +52,57 @@ impl ValidatedSchema {
     pub fn relation_mode(&self) -> datamodel_connector::RelationMode {
         self.relation_mode
     }
+
+    pub fn render_diagnostics(&self) -> String {
+        let mut out = Vec::new();
+
+        for error in self.diagnostics.errors() {
+            let (file_name, source, _) = &self.db[error.span().file_id];
+            error.pretty_print(&mut out, file_name, source.as_str()).unwrap();
+        }
+
+        String::from_utf8(out).unwrap()
+    }
 }
 
 /// The most general API for dealing with Prisma schemas. It accumulates what analysis and
 /// validation information it can, and returns it along with any error and warning diagnostics.
 pub fn validate(file: SourceFile, connectors: ConnectorRegistry<'_>) -> ValidatedSchema {
     let mut diagnostics = Diagnostics::new();
-    let db = ParserDatabase::new(file, &mut diagnostics);
-    let configuration = validate_configuration(db.ast(), &mut diagnostics, connectors);
+    let db = ParserDatabase::new_single_file(file, &mut diagnostics);
+    let configuration = validate_configuration(db.ast_assert_single(), &mut diagnostics, connectors);
+    let datasources = &configuration.datasources;
+    let out = validate::validate(db, datasources, configuration.preview_features(), diagnostics);
+
+    ValidatedSchema {
+        diagnostics: out.diagnostics,
+        configuration,
+        connector: out.connector,
+        db: out.db,
+        relation_mode: out.relation_mode,
+    }
+}
+
+/// The most general API for dealing with Prisma schemas. It accumulates what analysis and
+/// validation information it can, and returns it along with any error and warning diagnostics.
+pub fn validate_multi_file(files: Vec<(String, SourceFile)>, connectors: ConnectorRegistry<'_>) -> ValidatedSchema {
```

Comment:
Should `files` be `Vec<(&str, SourceFile)` instead?

Reply:
Not sure: in most places we will be getting that list, we will be getting `Vec<(String, SourceFile)>` and that won't coerce to `Vec<(&str, SourceFile)>` as easy as `&String` coerces to `&str`.